File: C++_Version/268Missing_Number.cpp
//Given an array containing n distinct numbers taken from 0, 1, 2, ..., n, find the one that is missing from the array.
//
//For example,
//Given nums = [0, 1, 3] return 2.
//
//Note:
//Your algorithm should run in linear runtime complexity.Could you implement it using only constant extra space complexity ?

#include <vector>
#include <iostream>
#include <algorithm>
#include <cstdlib>
#include <functional>
#include <string>
#include <unordered_map>
#include <map>
#include <set>
#include <unordered_set>
using namespace std;

class Solution {
public:
	int missingNumber(vector<int>& nums) {
		int sum = 0;
		int n = (int)nums.size();

		for (int i = 0; i < (int)nums.size(); ++i)
		{
			sum += nums[i];
		}

		return n * (n + 1) / 2 - sum;
	}
};

int main()
{
	Solution sol1;
	
	vector<int> nums1({ 1, 2, 3 });

	int receive = sol1.missingNumber(nums1);

	system("pause");

	return 0;
}

File: Java_Version/268_Missing_Number.java
/*
  Given an array containing n distinct numbers taken from 0, 1, 2, ..., n, find
the one that is missing from the array.

Example 1

        Input: [3,0,1]
        Output: 2
        Example 2

        Input: [9,6,4,2,3,5,7,0,1]
        Output: 8

Note:
  Your algorithm should run in linear runtime complexity. Could you implement
it using only constant extra space complexity?
*/

class Solution {

    public static void main(String[] args) {

        System.out.println(new Solution().missingNumber(new int[]{9,6,4,2,3,5,7,0,1}));
    }

    public int missingNumber(int[] nums) {

        if (nums == null || nums.length == 0) {
            throw new RuntimeException("illegal input");
        }

        int result = 0;
        for (int i = 0; i < nums.length; i++) {
            result ^= i;
            result ^= nums[i];
        }

        return (result ^ nums.length);
    }
}
